source/lac/scalapack.cc

```diff
@@ -323,117 +326,344 @@ void ScaLAPACKMatrix<NumberType>::invert()
 
 
 template <typename NumberType>
-std::vector<NumberType> ScaLAPACKMatrix<NumberType>::eigenvalues_symmetric()
+std::vector<NumberType> ScaLAPACKMatrix<NumberType>::eigenpairs_symmetric_by_index(const std::pair<unsigned int,unsigned int> &index_limits,
+    const bool compute_eigenvectors)
+{
+  // check validity of index limits
+  Assert (index_limits.first != numbers::invalid_unsigned_int,ExcMessage("index limits have invalid value"));
+  Assert (index_limits.second != numbers::invalid_unsigned_int,ExcMessage("index limits have invalid value"));
+  Assert (index_limits.first < (unsigned int)n_rows,ExcMessage("index limits have to be smaller than matrix size"));
+  Assert (index_limits.second < (unsigned int)n_rows,ExcMessage("index limits have to be smaller than matrix size"));
+
+  std::pair<unsigned int,unsigned int> idx = std::make_pair(std::min(index_limits.first,index_limits.second),
+                                                            std::max(index_limits.first,index_limits.second));
+
+  // compute all eigenvalues/eigenvectors
+  if (idx.first==0 && idx.second==(unsigned int)n_rows-1)
+    return eigenpairs_symmetric(compute_eigenvectors);
+  else
+    return eigenpairs_symmetric(compute_eigenvectors,idx);
+}
+
+
+
+template <typename NumberType>
+std::vector<NumberType> ScaLAPACKMatrix<NumberType>::eigenpairs_symmetric_by_value(const std::pair<NumberType,NumberType> &value_limits,
+    const bool compute_eigenvectors)
+{
+  Assert (!std::isnan(value_limits.first),ExcMessage("value limits have invalid value"));
+  Assert (!std::isnan(value_limits.second),ExcMessage("value limits have invalid value"));
+
+  std::pair<unsigned int,unsigned int> indices = std::make_pair(numbers::invalid_unsigned_int,numbers::invalid_unsigned_int);
+
+  return eigenpairs_symmetric(compute_eigenvectors,indices,value_limits);
+}
+
+
+
+template <typename NumberType>
+std::vector<NumberType>
+ScaLAPACKMatrix<NumberType>::eigenpairs_symmetric(const bool compute_eigenvectors,
+                                                  const std::pair<unsigned int, unsigned int> &eigenvalue_idx,
+                                                  const std::pair<NumberType,NumberType> &eigenvalue_limits)
 {
   Assert (state == LAPACKSupport::matrix,
           ExcMessage("Matrix has to be in Matrix state before calling this function."));
   Assert (property == LAPACKSupport::symmetric,
           ExcMessage("Matrix has to be symmetric for this operation."));
+
   Threads::Mutex::ScopedLock lock (mutex);
 
-  ScaLAPACKMatrix<NumberType> Z (grid->n_mpi_processes, grid, 1);
-  std::vector<NumberType> ev (n_rows);
+  bool valid_values = (std::isnan(eigenvalue_limits.first) || std::isnan(eigenvalue_limits.second)) ? false : true;
+  bool valid_indices = ((eigenvalue_idx.first==numbers::invalid_unsigned_int) || (eigenvalue_idx.second==numbers::invalid_unsigned_int)) ? false : true;
```

Comment:
you don't use this one (and their no need for it), so I would do 

1. rename: valid_values -> use_values
2. Add assert
```
if (!use_values)
  Assert (eigenvalue_idx.first!=numbers::invalid_unsigned_int && eigenvalue_idx.second!=numbers::invalid_unsigned_int, ExcInternalError())
```

Reply:
Actually, both are used. If both are `false` the routine `psyev` instead of `psyevx` is used to compute all eigenvalues/eigenvectors. Using `psyev` is safer to use, especially for eigenvector computation, as the workspace query for `psyevx` depends on a bunch of parameters (e.g. eigenvalue distribution) and might cause ScaLAPACK to run out of memory.